File: server/app/api/v1/routes/question_packs.py
from math import ceil
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import func, or_
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload
from typing import List

from app.api.v1.deps import get_current_user, get_session, require_role
from app.models.user import User, UserRole
from app.models.enums import PackStatus
from app.models.question_pack import QuestionPack
from app.models.question import Question
from app.core.pagination import get_pagination_bounds
from app.schemas.question_pack import QuestionPack as QuestionPackSchema, QuestionPackCreate, QuestionCreate, Question as QuestionSchema
from app.services.question_pack_service import QuestionPackService

router = APIRouter(tags=["question-packs"])

# ── Static question routes MUST come before /{pack_id} to avoid routing conflicts ──

@router.get("/questions", response_model=List[QuestionSchema])
async def list_all_questions(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_session)
):
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    service = QuestionPackService(db)
    return await service.get_all_questions(skip=skip, limit=limit)


@router.get("/questions/paged")
async def list_all_questions_paged(
    page: int = 1,
    page_size: int = 10,
    search: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_session)
):
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(status_code=403, detail="Insufficient permissions")

    offset, normalized_page, normalized_page_size = get_pagination_bounds(page=page, page_size=page_size)

    base_query = select(Question)

    if search:
        search_term = f"%{search.strip()}%"
        if search_term != "%%":
            base_query = base_query.where(func.lower(Question.question_text).like(func.lower(search_term)))

    total_stmt = select(func.count()).select_from(base_query.subquery())
    total = (await db.execute(total_stmt)).scalar_one()

    result = await db.execute(
        base_query.order_by(Question.created_at.desc()).offset(offset).limit(normalized_page_size)
    )
    items = result.scalars().all()

    return {
        "items": items,
        "total": total,
        "page": normalized_page,
        "page_size": normalized_page_size,
        "total_pages": ceil(total / normalized_page_size) if total else 1,
    }

@router.patch("/questions/{question_id}", response_model=QuestionSchema)
async def update_question(
    question_id: str,
    question_in: QuestionCreate,
    current_user: User = Depends(require_role(UserRole.ADMIN)),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    updated = await service.update_question(question_id, question_in.model_dump(exclude_unset=True))
    if not updated:
        raise HTTPException(status_code=404, detail="Question not found")
    return updated

@router.delete("/questions/{question_id}")
async def delete_question(
    question_id: str,
    current_user: User = Depends(require_role(UserRole.ADMIN)),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    success = await service.delete_question(question_id)
    if not success:
        raise HTTPException(status_code=404, detail="Question not found")
    return {"status": "success"}

# ── Pack-level routes ──

@router.get("/", response_model=List[QuestionPackSchema])
async def list_question_packs(
    skip: int = 0,
    limit: int = 100,
    domain: str = None,
    status: str = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    filters = {}
    if domain: filters["domain"] = domain
    if current_user.role != UserRole.ADMIN:
        filters["status"] = PackStatus.PUBLISHED.value
    elif status:
        filters["status"] = status
    return await service.get_all(skip=skip, limit=limit, filters=filters)


@router.get("/paged")
async def list_question_packs_paged(
    page: int = 1,
    page_size: int = 10,
    search: Optional[str] = None,
    domain: Optional[str] = None,
    status: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_session)
):
    offset, normalized_page, normalized_page_size = get_pagination_bounds(page=page, page_size=page_size)

    base_query = select(QuestionPack)

    if domain:
        base_query = base_query.where(QuestionPack.domain == domain)

    if current_user.role != UserRole.ADMIN:
        base_query = base_query.where(QuestionPack.status == PackStatus.PUBLISHED.value)
    elif status:
        base_query = base_query.where(QuestionPack.status == status)

    if search:
        search_term = f"%{search.strip()}%"
        if search_term != "%%":
            base_query = base_query.where(
                or_(
                    func.lower(QuestionPack.pack_name).like(func.lower(search_term)),
                    func.lower(QuestionPack.topic).like(func.lower(search_term)),
                    func.lower(QuestionPack.description).like(func.lower(search_term)),
                )
            )

    total_stmt = select(func.count()).select_from(base_query.subquery())
    total = (await db.execute(total_stmt)).scalar_one()

    result = await db.execute(
        base_query
        .order_by(QuestionPack.created_at.desc())
        .offset(offset)
        .limit(normalized_page_size)
        .options(selectinload(QuestionPack.questions))
    )
    items = result.scalars().all()

    return {
        "items": items,
        "total": total,
        "page": normalized_page,
        "page_size": normalized_page_size,
        "total_pages": ceil(total / normalized_page_size) if total else 1,
    }

@router.post("/", response_model=QuestionPackSchema)
async def create_question_pack(
    pack_in: QuestionPackCreate,
    current_user: User = Depends(require_role(UserRole.ADMIN)),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    return await service.create_pack(pack_in, current_user.id)

@router.get("/{pack_id}", response_model=QuestionPackSchema)
async def get_question_pack(
    pack_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    pack = await service.get_pack(pack_id)
    if not pack:
        raise HTTPException(status_code=404, detail="Question pack not found")
    pack_status = pack.status.value if hasattr(pack.status, "value") else pack.status
    if current_user.role != UserRole.ADMIN and pack_status != PackStatus.PUBLISHED.value:
        raise HTTPException(status_code=404, detail="Question pack not found")
    return pack

@router.patch("/{pack_id}", response_model=QuestionPackSchema)
async def update_question_pack(
    pack_id: str,
    pack_in: QuestionPackCreate,
    current_user: User = Depends(require_role(UserRole.ADMIN)),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    updated = await service.update_pack(pack_id, pack_in.model_dump(exclude_unset=True))
    if not updated:
        raise HTTPException(status_code=404, detail="Question pack not found")
    return updated

@router.delete("/{pack_id}")
async def delete_question_pack(
    pack_id: str,
    current_user: User = Depends(require_role(UserRole.ADMIN)),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    success = await service.delete_pack(pack_id)
    if not success:
        raise HTTPException(status_code=404, detail="Question pack not found")
    return {"status": "success"}

@router.post("/{pack_id}/questions", response_model=QuestionSchema)
async def add_question_to_pack(
    pack_id: str,
    question_in: QuestionCreate,
    current_user: User = Depends(require_role(UserRole.ADMIN)),
    db: AsyncSession = Depends(get_session)
):
    service = QuestionPackService(db)
    return await service.add_question(pack_id, question_in)

File: client/src/lib/api/questionPacks.ts
import http from './http'
import { SkillDomain } from './skills'
import { Difficulty } from './tests'

export enum PackStatus {
    DRAFT = "DRAFT",
    PUBLISHED = "PUBLISHED",
    ARCHIVED = "ARCHIVED",
}

export enum QuestionOption {
    A = "A",
    B = "B",
    C = "C",
    D = "D",
}

export interface Question {
    id: string;
    pack_id: string;
    question_text: string;
    option_a: string;
    option_b: string;
    option_c: string;
    option_d: string;
    correct_option: QuestionOption;
    explanation?: string;
    created_at: string;
}

export interface QuestionPack {
    id: string;
    pack_name: string;
    domain: SkillDomain;
    topic?: string;
    difficulty: Difficulty;
    description?: string;
    status: PackStatus;
    created_by_id: string;
    created_at: string;
    published_at?: string;
    questions: Question[];
}

export interface QuestionPackCreate {
    pack_name: string;
    domain: SkillDomain;
    topic?: string;
    difficulty?: Difficulty;
    description?: string;
    status?: PackStatus;
}

export interface QuestionCreate {
    question_text: string;
    option_a: string;
    option_b: string;
    option_c: string;
    option_d: string;
    correct_option: QuestionOption;
    explanation?: string;
}

export interface PaginatedQuestionPacks {
    items: QuestionPack[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
}

export interface PaginatedQuestions {
    items: Question[];
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
}

export const questionPacksApi = {
    listPacks: async (skip: number = 0, limit: number = 100, domain?: string, status?: string): Promise<QuestionPack[]> => {
        const response = await http.get<QuestionPack[]>('/api/v1/question-packs/', {
            params: { skip, limit, domain, status }
        });
        return response.data;
    },

    listPacksPaginated: async (params?: {
        page?: number;
        pageSize?: number;
        search?: string;
        domain?: string;
        status?: string;
    }): Promise<PaginatedQuestionPacks> => {
        const response = await http.get<PaginatedQuestionPacks>('/api/v1/question-packs/paged', {
            params: {
                ...(params?.page ? { page: params.page } : {}),
                ...(params?.pageSize ? { page_size: params.pageSize } : {}),
                ...(params?.search ? { search: params.search } : {}),
                ...(params?.domain ? { domain: params.domain } : {}),
                ...(params?.status ? { status: params.status } : {}),
            }
        });
        return response.data;
    },

    getPack: async (id: string): Promise<QuestionPack> => {
        const response = await http.get<QuestionPack>(`/api/v1/question-packs/${id}`);
        return response.data;
    },

    createPack: async (packData: QuestionPackCreate): Promise<QuestionPack> => {
        const response = await http.post<QuestionPack>('/api/v1/question-packs/', packData);
        return response.data;
    },

    updatePack: async (id: string, packData: Partial<QuestionPackCreate>): Promise<QuestionPack> => {
        const response = await http.patch<QuestionPack>(`/api/v1/question-packs/${id}`, packData);
        return response.data;
    },

    deletePack: async (id: string): Promise<{ status: string }> => {
        const response = await http.delete<{ status: string }>(`/api/v1/question-packs/${id}`);
        return response.data;
    },

    addQuestion: async (packId: string, questionData: QuestionCreate): Promise<Question> => {
        const response = await http.post<Question>(`/api/v1/question-packs/${packId}/questions`, questionData);
        return response.data;
    },

    listAllQuestions: async (skip: number = 0, limit: number = 100): Promise<Question[]> => {
        const response = await http.get<Question[]>('/api/v1/question-packs/questions', {
            params: { skip, limit }
        });
        return response.data;
    },

    listAllQuestionsPaginated: async (params?: { page?: number; pageSize?: number; search?: string }): Promise<PaginatedQuestions> => {
        const response = await http.get<PaginatedQuestions>('/api/v1/question-packs/questions/paged', {
            params: {
                ...(params?.page ? { page: params.page } : {}),
                ...(params?.pageSize ? { page_size: params.pageSize } : {}),
                ...(params?.search ? { search: params.search } : {}),
            }
        });
        return response.data;
    },

    updateQuestion: async (id: string, questionData: Partial<QuestionCreate>): Promise<Question> => {
        const response = await http.patch<Question>(`/api/v1/question-packs/questions/${id}`, questionData);
        return response.data;
    },

    deleteQuestion: async (id: string): Promise<{ status: string }> => {
        const response = await http.delete<{ status: string }>(`/api/v1/question-packs/questions/${id}`);
        return response.data;
    }
}
